partitioned-heat-conduction/solver-fenics/heat_pySDC.py

```diff
@@ -0,0 +1,346 @@
+# Setup pySDC controller
+##################################################################
+'''
+pySDC can be installed via `pip install pySDC`.
+If you want to use the developer version, the source code repository can be cloned from "https://github.com/Parallel-in-Time/pySDC".
+For more information on pySDC, see also "https://parallel-in-time.org/pySDC/"
+'''
+
+# initialize level parameters
+level_params = dict()
+level_params['restol'] = 1e-11
+level_params['dt'] = pySDC_dt
+
+# initialize step parameters
+step_params = dict()
+step_params['maxiter'] = 40
+
+# initialize sweeper parameters
+sweeper_params = dict()
+sweeper_params['quad_type'] = 'LOBATTO'
+sweeper_params['num_nodes'] = 4
+
+# initialize problem parameters
+problem_params = dict()
+problem_params['mesh'] = domain_mesh
+problem_params['function_space'] = V
+problem_params['coupling_boundary'] = coupling_boundary
+problem_params['remaining_boundary'] = remaining_boundary
+problem_params['solution_expr'] = u_D
+problem_params['forcing_term_expr'] = forcing_expr
+problem_params['precice_ref'] = precice
+problem_params['coupling_expr'] = coupling_expression
+problem_params['participant_name'] = participant_name
+
+# initialize controller parameters
+controller_params = dict()
+controller_params['logger_level'] = 30
+
+# fill description dictionary for easy step instantiation
+description = dict()
+description['problem_class'] = fenics_heat_2d
+description['problem_params'] = problem_params
+description['sweeper_class'] = imex_1st_order_mass
+description['sweeper_params'] = sweeper_params
+description['level_params'] = level_params
+description['step_params'] = step_params
+
+# Controller for time stepping
+controller = controller_nonMPI(num_procs=1, controller_params=controller_params, description=description)
+
+# Reference to problem class for easy access to exact solution
+P = controller.MS[0].levels[0].prob
+
+##################################################################
+
+    
+
+#################################################################
+# Coupling loop
+#################################################################
+
+while precice.is_coupling_ongoing():
+
+    # write checkpoint
+    if precice.requires_writing_checkpoint():
+        precice.store_checkpoint(u_n, t, n)
+
+        # output solution and reference solution at t_n+1 and substeps (read from buffer)
+        print('output u^%d and u_ref^%d' % (n, n))
+        for sample in u_write:
+            temperature_out << sample
+
+        for sample in ref_write:
+            ref_out << sample
+
+        for sample in error_write:
+            error_out << error_pointwise
+
+    precice_dt = precice.get_max_time_step_size()
+    dt.assign(np.min([pySDC_dt, precice_dt]))
+
+    # Update start time of the current preCICE time step within
+    # the problem class.
+    # This is necessary, that the coupling expression update in the problem class
+    # is executed correctly.
+    P.set_t_start(t)
```

Comment:
Ok. Here `P` is used. I would actually consider writing the following instead:

```suggestion
    controller.MS[0].levels[0].prob.set_t_start(t)
```

Yes. This is a bit more lengthy but it saves you one indirection and clarifies the connection to the controller.